ntoskrnl/inbv/inbv.c

```diff
@@ -932,18 +953,145 @@ DisplayBootBitmap(IN BOOLEAN TextMode)
     }
 
 #ifdef INBV_ROTBAR_IMPLEMENTED
+    if (RotBarSelection != RB_PROGRESS_BAR)
+    {
+        /* Hide simple progress bar if not used */
+        ShowProgressBar = FALSE;
+    }
+
     /* Do we have a system thread? */
-    if (RotBarThreadActive)
+    if (!RotBarThreadActive)
     {
-        /* We do, initialize the progress bar */
+        /* We don't, initialize the progress bar */
         InbvAcquireLock();
         RotBarSelection = TempRotBarSelection;
-        // InbvRotBarInit();
         InbvReleaseLock();
+        /* Start progress bar thread */
+        InbvRotBarInit();
     }
 #endif
 }
 
+#ifdef INBV_ROTBAR_IMPLEMENTED
+static
+VOID
+NTAPI
+InbvRotationThread(
+    _In_ PVOID Context)
+{
+    ULONG X, Y, Index = 0, Total = 18;
+
+    InbvAcquireLock();
+    X = ProgressBarLeft + 2;
+    Y = ProgressBarTop + 2;
+    InbvReleaseLock();
+
+    while (TRUE)
```

Comment:
You should use a (global, or passed from `Context`) flag here instead, so that in your InbvRotBarStop() function you set this flag to FALSE (for example) and then wait on the thread to finish.

Reply:
It uses check `RotBarSelection != RB_UNSPECIFIED` but inside a lock.

Reply:
I just thought much more about that, and a possible much more simpler stop condition would be checking whether InbvCheckDisplayOwnership() becoming FALSE. In which case this would simplify much more your code, and potentially allowing you not to care about stopping the thread later as it would stop by itself...